### PR TITLE
feat(mt-438): support for li token auth guard

### PR DESCRIPTION
**Pull request by chintan-liveintent**

### CODE REVIEW

File: src/Providers/LIAuthServiceProvider.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+            return new RequestGuard(function ($request) use ($config) {
+                return (new LITokenGuard(
+                    Auth::createUserProvider($config['provider']),
+                ))->user($request);
+            }, $app['request']);
+        });
+
+        // Causes issues with an infinite loop, cannot set `li_token` as default guard with current implementation
+//        Auth::shouldUse('li_token');
```
Comment (chintan-liveintent) [inline src/Providers/LIAuthServiceProvider.php]: When I attempted to use this to set the default auth guard, (or even manually configuring the file and putting `'li_token'` as the default) this resulted in an infinite loop of authenticating the token. Perhaps there's something wrong with the way RequestGuards work. 

I was going to change `LITokenGuard` into an actual Guard object, but figured I'd push this up as is for now. Let me know if you want me to try that.

File: config/liveintent.php [hunk 2]
```diff
@@ -41,11 +41,6 @@
     |
     */
     'auth' => [
-
-        'li_token' => [
-            'public_key' => env('LI_TOKEN_PUBLIC_KEY')
-        ]
-
```
Comment (tam5) [inline config/liveintent.php]: it looks like we are removing this in favor of doing it here: https://github.com/LiveIntent/laravel-common/pull/1/files#diff-64af6efa8d129988758fd1a3f84d368ba5242c179edc2d1e5b2cdd26a92dbeb1R31-R33 

can you explain why the change?

File: src/Providers/LIAuthServiceProvider.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        if (config('auth.li_token.keys.public') == null) {
+            config(['auth.li_token.keys.public' => env('TESSELLATE_PUBLIC_KEY') ]);
+        }
+
+        if (config('auth.guards.li_token') == null) {
+            config(['auth.guards.li_token' => [
+                'driver' => 'li_token',
+                'provider' => 'users',
+            ]]);
+        }
```
Comment (tam5) [inline src/Providers/LIAuthServiceProvider.php]: I'm not sure i fully understand what is being done here. It looks like maybe you are trying to protect against a case where a consumer has a different config set up in their app maybe? Would appreciate clarification before I comment on it

File: src/Providers/LIAuthServiceProvider.php (hunk 1 above)
Comment (tam5) [reply src/Providers/LIAuthServiceProvider.php]: This package should only expose the guard as being available to use, but it should not set the default guard in the consuming app IMO. Remember, this package is meant to provide a set of tools that can be used when desired, but we want to make sure we give the flexibility and choice to the consumer and don't introduce too many hidden "behavior" changes to consuming apps.

To set the default guard I would expect a consuming app to set the value here https://github.com/LiveIntent/tpl-laravel/blob/main/config/auth.php#L17 to li_token.

File: src/Providers/LIAuthServiceProvider.php [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+use Illuminate\Support\Facades\Auth;
+use LiveIntent\LaravelCommon\Auth\Guards\LITokenGuard;
+use Illuminate\Foundation\Support\Providers\AuthServiceProvider as ServiceProvider;
+
+class LIAuthServiceProvider extends ServiceProvider
+{
+    public function __construct($app)
+    {
+        parent::__construct($app);
+        $this->registerEnvironmentVariables();
```
Comment (tam5) [inline src/Providers/LIAuthServiceProvider.php]: Waiting on your clarification to better understand what's intended here

File: src/LaravelCommon.php [hunk 5]
```diff
@@ -2,12 +2,24 @@
...
+use LiveIntent\LaravelCommon\Providers\LIAuthServiceProvider;
 
 class LaravelCommon
 {
+    /**
+     * Register Tessellate token authentication guard
+     */
+    public static function authGuard()
+    {
+        App::register(AuthServiceProvider::class);
```
Comment (tam5) [inline src/LaravelCommon.php]: Can you explain why we'd prefer to register the Illuminate\Auth\AuthServiceProvider here rather than leaving the default/standard approach as is?

File: src/Providers/LIAuthServiceProvider.php [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        if (config('auth.guards.li_token') == null) {
+            config(['auth.guards.li_token' => [
+                'driver' => 'li_token',
+                'provider' => 'users',
+            ]]);
+        }
+
+        $this->policies = config('auth.li_token.policies', []);
```
Comment (tam5) [inline src/Providers/LIAuthServiceProvider.php]: Same recurring question I guess, why here have we introduced a new way of defining the policies?

File: src/Providers/LIAuthServiceProvider.php (hunk 1 above)
Comment (chintan-liveintent) [reply src/Providers/LIAuthServiceProvider.php]: Sure, that makes sense. This line checks if the default guard is set, and if it isn't, only then will it set it to `li_token`. However, setting the default guard through that mechanism or as seen in the `auth.php` on line 17 (as per your link) ends up causing an infinite loop even though it recognizes that there is only one guard.

File: src/Providers/LIAuthServiceProvider.php (hunk 6 above)
Comment (chintan-liveintent) [reply src/Providers/LIAuthServiceProvider.php]: This was done so that the `AuthServiceProvider` didn't need to be extended in order to add support for policies in Liam. If we want to let it continue to require the class to be extended, I can revert this change

File: src/LaravelCommon.php (hunk 5 above)
Comment (chintan-liveintent) [reply src/LaravelCommon.php]: Sure, so I was trying to determine where we want to draw the line between adding a one function call to support adding a default auth to any service versus calling a function and still following convention by adding the dependency in the `Kernel.php` file.

File: src/Providers/LIAuthServiceProvider.php (hunk 3 above)
Comment (chintan-liveintent) [reply src/Providers/LIAuthServiceProvider.php]: Same thing with this one -- this will automatically configure the `li_token` auth guard driver and provider if one is not specified already. If for whatever reason, a different provider is wanted, it can be overridden.

Another way we would support the "one function to enable auth" mechanism.

Another approach we could take here is to put this in the `liveintent.php` file and here we would load from that instead. Then we'd need all projects using laravel-common to update that file. This way it would be more consistent with expectations and less "magic". What do you think?

File: config/liveintent.php (hunk 2 above)
Comment (chintan-liveintent) [reply config/liveintent.php]: Yep same reasoning as mentioned in this comment: https://github.com/LiveIntent/laravel-common/pull/1/files#r824009819

**Comment by chintan-liveintent** [comment]
@tam5 With this first pass, I tried to push this to one extreme of supporting the conventional way of allowing auth through importing the dependencies within `kernel.php` versus calling one function in `LaravelCommon` to enable auth.

Here is where I think there were some places where we may need to make changes (as per our discussions on this PR):
1. Do / do not change the way the default configuration values are set to be done through either `auth.php` and/or `liveintent.php`
2. Do / do not auto-register `Illuminate\Auth\AuthServiceProvider` and instead expect it to exist when `LaravelCommon->authGuard()` is called
3. Do / do not support auth policies through configuration, and instead require that it be added by services through extending `LIAuthServiceProvider` (this would mean that `LaravelCommon->authGuard()` would not be used in this circumstance)
4. Still need to determine how to set it as the default guard as creating it as seen by `LarvelCommon::registerCustomGuards()` doesn't seem to allow us to set the default driver to `li_token` still.


Let me know where you'd like the line drawn between current conventions and simplicity in usage

File: src/LaravelCommon.php [hunk 7]
```diff
@@ -2,12 +2,32 @@
...
 
 class LaravelCommon
 {
+    /**
+     * Register Tessellate token authentication guard
+     */
+    public static function authGuard()
+    {
+        Auth::extend('li_token', function ($app, $_name, array $config) {
+            throw_if($config['provider'] == null);
```
Comment (jiechaowang) [inline src/LaravelCommon.php]: looks fancy and expressive but because you didn't specify the second argument, I wonder if it will provide us enough info when it throws an exception?

File: src/LaravelCommon.php [hunk 8]
```diff
@@ -2,12 +2,32 @@
...
+
+            return new RequestGuard(function ($request) use ($config) {
+                return (new LITokenGuard(
+                    Auth::createUserProvider($config['provider']),
+                    config('auth.li_token.keys.public') ?? config('liveintent.auth.li_token.keys.public')
```
Comment (jiechaowang) [inline src/LaravelCommon.php]: Not sure why do this, can we make sure it is always either `auth.li_token.keys.public` or always `liveintent.auth.li_token.keys.public`

File: src/LaravelCommon.php [hunk 9]
```diff
@@ -2,12 +2,32 @@
...
 use Illuminate\Support\Facades\Route;
 use Illuminate\Foundation\Http\Events\RequestHandled;
+use LiveIntent\LaravelCommon\Auth\Guards\LITokenGuard;
 
 class LaravelCommon
 {
+    /**
+     * Register Tessellate token authentication guard
+     */
+    public static function authGuard()
```
Comment (jiechaowang) [inline src/LaravelCommon.php]: I wonder what you guys think, but personally I think the function name `registerAuthGuard()` conveys the its purpose better, yeah I know it's longer

File: src/LaravelCommon.php (hunk 8 above)
Comment (tam5) [reply src/LaravelCommon.php]: I do agree with @jiechaowang here, I would stick to one approach, and probably I'd prefer to keep the current approach of the liveintent file.

Side note, the config function allows a second parameter for a default value. Just another reason why I feel like this looks a bit "off" to me.

File: src/LaravelCommon.php (hunk 9 above)
Comment (tam5) [reply src/LaravelCommon.php]: yeah I went back/forth with that, I think i agree that register is more explicit and might be more clear

File: src/LaravelCommon.php (hunk 7 above)
Comment (tam5) [reply src/LaravelCommon.php]: I'd actually probably leave this throw out honestly, I think it's not really needed and may actually hide the real problem that we'd otherwise be able to determine from the error message that was thrown. (Also this very well could be part of the infinite loop problem :) )

File: config/liveintent.php [hunk 10]
```diff
@@ -41,11 +41,11 @@
     |
     */
     'auth' => [
-
         'li_token' => [
-            'public_key' => env('LI_TOKEN_PUBLIC_KEY')
+            'keys' => [
+                'public' => env('TESSELLATE_PUBLIC_KEY')
+            ]
```
Comment (tam5) [inline config/liveintent.php]: any particular reason for the structural change?

File: src/Auth/Guards/LITokenGuard.php [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Create a new internal token guard instance.
+     * @throws Throwable
+     */
+    public function __construct(
+        private UserProvider $userProvider,
+        $plaintextPublicKey,
+    ) {
+        throw_if($plaintextPublicKey == null);
```
Comment (tam5) [inline src/Auth/Guards/LITokenGuard.php]: 1) should we just type hint the key as a string so it cannot be null anyway?
2) maybe we don't really need to accept the key in the constructor? then we might be able to make the code shorter (i.e. configuration object is only ever used inside the user function, rather than created in one place and used in another

File: src/LaravelCommon.php (hunk 7 above)
Comment (chintan-liveintent) [reply src/LaravelCommon.php]: Sure, will remove this piece. I think you may be right that this may be the culprit for the infinite loop, but will test that piece.

File: src/LaravelCommon.php (hunk 8 above)
Comment (chintan-liveintent) [reply src/LaravelCommon.php]: Yeah, I was on the fence on whether or not I wanted to do this--originally I was thinking if there was a reason to override it, but keep the liveintent config file untouched. However, I think you guys are right that it should just work one way.

File: config/liveintent.php (hunk 10 above)
Comment (chintan-liveintent) [reply config/liveintent.php]: Nope, just realized I had the wrong structure in the original commit. I'm fine with reverting this to the way it was previously.

File: src/Auth/Guards/LITokenGuard.php (hunk 11 above)
Comment (chintan-liveintent) [reply src/Auth/Guards/LITokenGuard.php]: 1. Yeah, that makes sense. Wasn't sure if the better approach to null handling was to check within the function.

2. I was thinking that LITokenGuard shouldn't care how the public key is loaded. It would make this class a bit easier to write test for/with if that's needed somewhere without really needing to understand the inner workings of this class. However, I think to facilitate the ease of use and it working "out of the box", it makes it a bit easier to not have it as a constructor value. I'm fine with either approach if you prefer one way over another for consistency. I'll remove it from the constructor since we don't have any tests for this anyway right now.

Comment (tam5) [reply src/Auth/Guards/LITokenGuard.php]: There are nice ways to test this even without the key being passed through the constructor, since it's still passed through the config, I can show you.

**Review by tam5 — APPROVED** [review]
